### DISCUSSION TITLE
ADC sample rate and resolution

### DISCUSSION THREAD

**Discussion by JyeSmith** [EdgeTX/edgetx#3492] [General]
Hey ETX!

This topic has been raised again with ELRS and BF, but it will be much better to have everyone talking/collaborating in a single place.

From my side of things ELRS has the capability to send 12b channels OTA for using FLRC.  This could be beneficial for BF, feed forward, racing, etc.

Im not an expert on these things so Ill let BF devs chime in on the details and what will be optimal. @ctzsnooze

@rotorman youve already shared some great info and links with me on how ETX current operates and potential plans.  Do you mind rehashing it for others?

**Comment by ctzsnooze** [comment]
Yes I think it would be great to get the absolute best possible data path.  Happy to help if I can.

**Reply by rotorman** [reply]
Current EdgeTX codebase samples analog inputs with 12-bit resolution ( https://github.com/EdgeTX/edgetx/pull/3488 ), but internally uses 11-bits for calculations. Plans are afoot to upgrade internal calculations to 16-bits, but no code or POC has been done to the best of my knowledge in this direction yet.

From timing perspective, sampling once all analogs, e.g. on TX16S, takes ca. 25 µs. Present code uses 4-fold oversampling:
https://github.com/EdgeTX/edgetx/blob/54e0d350e8ad3cfef3382cfe6267a19863887ff2/radio/src/hal/adc_driver.cpp#L150-L166
taking on TX16S ca. 100 µs to sample all the analog channels (here toggling a pin during ADC sampling and conversion and measuring this with a logic analyzer):
![grafik](https://user-images.githubusercontent.com/21011587/232981735-d6fcef05-2a8d-49dd-8ae5-78158b5aa089.png)

Regarding elevating the overall sampling higher than 1 kHz, please consider the computational load increase due to more often required calculations of the mixer task. At present I have no immediate statistics to present how long a computational cycle would take. Also to consider is that EdgeTX supported radios use a small variety of STM32F2 and STM32F4 MCUs with differing computational performance.

**Reply by ctzsnooze** [reply]
Thank you for the explanation.
Am I correct in understanding that, for any given gimbal channel, four samples are taken, at approximately 25us intervals, and then those four samples are averaged, with analog decimation to an 11 bit value?
How often is the `adcRead()` task called? Is it synchronised to the rC link frequency, ie, called prior to the time of transmission, or is data collection and data transmission asynchronous?
If the ADC filter is disabled, is there any filtering of `adcValues[x]` before being transmitted?

**Reply by rotorman** [reply]
The ADC sampling is synchronized with RF-modules that support dynamic synchronisation, such as ExpressLRS. With ELRS configured at 1000 Hz packet rate, `adcRead()` is called 1000 times per second. `adcRead()` gets called from `getADC()`: https://github.com/EdgeTX/edgetx/blob/54e0d350e8ad3cfef3382cfe6267a19863887ff2/radio/src/hal/adc_driver.cpp#L250-L264

`getADC()` gets called by `doMixerCalculations()`:
https://github.com/EdgeTX/edgetx/blob/54e0d350e8ad3cfef3382cfe6267a19863887ff2/radio/src/tasks/mixer_task.cpp#L223-L244

`doMixerCalculations()` is called in an endless loop of a mixer task (=FreeRTOS task):
https://github.com/EdgeTX/edgetx/blob/54e0d350e8ad3cfef3382cfe6267a19863887ff2/radio/src/tasks/mixer_task.cpp#L141-L193

The 25 µs period samples all analog channels of the TX16S. They are initially sampled with 12-bit precision.

ADC filtering is done on top of the above 4-fold sampling and averaging. More about it in: https://github.com/EdgeTX/edgetx/blob/54e0d350e8ad3cfef3382cfe6267a19863887ff2/radio/src/hal/adc_driver.cpp#L285-L362

**Comment by JBKingdon** [comment]
Where does start_conversion get implemented?

**Reply by rotorman** [reply]
`start_conversion` is a function pointer of `etx_hal_adc_driver_t` struct:
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/hal/adc_driver.h#L53-L57
that gets passed to `adcInit()`:
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/hal/adc_driver.cpp#L99-L113
`adcInit()`gets called in `boardInit()`, here for landscape orientation color radios:
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/targets/horus/board.cpp#L167
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/targets/horus/board.cpp#L40
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/targets/common/arm/stm32/stm32_hal_adc.cpp#L470-L474
`start_conversion` is mapped to `stm32_hal_adc_start_read()`:
https://github.com/EdgeTX/edgetx/blob/8b40fb771ac8f4005d94083f950ed0ac9a593c02/radio/src/targets/common/arm/stm32/stm32_hal_adc.cpp#L411-L431

**Reply by JBKingdon** [reply]
Awesome, thanks! That would have taken me ages to track down :)
